File: tablet_qt/tasks/smast.cpp
/*
    Copyright (C) 2012-2018 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.
*/

#include "smast.h"
#include "maths/mathfunc.h"
#include "lib/stringfunc.h"
#include "questionnairelib/commonoptions.h"
#include "questionnairelib/namevaluepair.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/qumcqgrid.h"
#include "questionnairelib/qutext.h"
#include "tasklib/taskfactory.h"
using mathfunc::noneNull;
using mathfunc::totalScorePhrase;
using stringfunc::bold;
using stringfunc::strnum;
using stringfunc::strseq;

const int FIRST_Q = 1;
const int N_QUESTIONS = 13;
const int MAX_SCORE = N_QUESTIONS;
const QString QPREFIX("q");

const QVector<int> REVERSE_SCORED_Q{1, 4, 5};

const QString Smast::SMAST_TABLENAME("smast");


void initializeSmast(TaskFactory& factory)
{
    static TaskRegistrar<Smast> registered(factory);
}


Smast::Smast(CamcopsApp& app, DatabaseManager& db, const int load_pk) :
    Task(app, db, SMAST_TABLENAME, false, false, false)  // ... anon, clin, resp
{
    addFields(strseq(QPREFIX, FIRST_Q, N_QUESTIONS), QVariant::String);

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.
}


// ============================================================================
// Class info
// ============================================================================

QString Smast::shortname() const
{
    return "SMAST";
}


QString Smast::longname() const
{
    return tr("Short Michigan Alcohol Screening Test");
}


QString Smast::menusubtitle() const
{
    return tr("13-item Y/N self-report scale.");
}


QString Smast::infoFilenameStem() const
{
    return "mast";
}


// ============================================================================
// Instance info
// ============================================================================

bool Smast::isComplete() const
{
    return noneNull(values(strseq(QPREFIX, FIRST_Q, N_QUESTIONS)));
}


QStringList Smast::summary() const
{
    return QStringList{totalScorePhrase(totalScore(), MAX_SCORE)};
}


QStringList Smast::detail() const
{
    const int total = totalScore();
    const QString likelihood =
            total >= 3 ? xstring("problem_probable")
                       : (total >= 2 ? xstring("problem_possible")
                                     : xstring("problem_unlikely"));
    const QString scores = ", " + xstring("scores") + " ";

    QStringList lines = completenessInfo();
    for (int q = FIRST_Q; q <= N_QUESTIONS; ++q) {
        lines.append(fieldSummary(strnum(QPREFIX, q),
                                  xstring(strnum("q", q, "_s")),
                                  " ") +
                     scores + bold(QString::number(score(q))));
    }
    lines.append("");
    lines += summary();
    lines.append("");
    lines.append(xstring("problem_likelihood") + " " + bold(likelihood));
    return lines;
}


OpenableWidget* Smast::editor(const bool read_only)
{
    QVector<QuestionWithOneField> qfields;
    for (int i = FIRST_Q; i <= N_QUESTIONS; ++i) {
        qfields.append(QuestionWithOneField(xstring(strnum("q", i)),
                                            fieldRef(strnum(QPREFIX, i))));
    }
    const QVector<McqGridSubtitle> sub{
        {5, ""},
        {10, ""},
        {15, ""},
    };

    QuPagePtr page((new QuPage{
        new QuText(xstring("stem")),
        (new QuMcqGrid(qfields, CommonOptions::yesNoChar()))->setSubtitles(sub),
    })->setTitle(xstring("title")));

    Questionnaire* questionnaire = new Questionnaire(m_app, {page});
    questionnaire->setType(QuPage::PageType::Patient);
    questionnaire->setReadOnly(read_only);
    return questionnaire;
}


// ============================================================================
// Task-specific calculations
// ============================================================================

int Smast::score(const int question) const
{
    const QVariant v = value(strnum(QPREFIX, question));
    if (v.isNull()) {
        return 0;  // to avoid silly scoring of incomplete tasks
    }
    const bool yes = v.toString() == CommonOptions::YES_CHAR;
    if (REVERSE_SCORED_Q.contains(question)) {
        return yes ? 0 : 1;
    } else {
        return yes ? 1 : 0;
    }
}


int Smast::totalScore() const
{
    int total = 0;
    for (int q = FIRST_Q; q <= N_QUESTIONS; ++q) {
        total += score(q);
    }
    return total;
}

File: server/camcops_server/tasks/smast.py
#!/usr/bin/env python
# camcops_server/tasks/smast.py

"""
===============================================================================

    Copyright (C) 2012-2018 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.

===============================================================================
"""

from typing import Any, Dict, List, Tuple, Type

from cardinal_pythonlib.stringfunc import strseq
from sqlalchemy.ext.declarative import DeclarativeMeta
from sqlalchemy.sql.sqltypes import Integer

from camcops_server.cc_modules.cc_constants import CssClass
from camcops_server.cc_modules.cc_ctvinfo import CtvInfo, CTV_INCOMPLETE
from camcops_server.cc_modules.cc_db import add_multiple_columns
from camcops_server.cc_modules.cc_html import answer, tr, tr_qa
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_sqla_coltypes import (
    CharColType,
    SummaryCategoryColType,
)
from camcops_server.cc_modules.cc_summaryelement import SummaryElement
from camcops_server.cc_modules.cc_task import (
    get_from_dict,
    Task,
    TaskHasPatientMixin,
)
from camcops_server.cc_modules.cc_trackerhelpers import (
    TrackerLabel,
    TrackerInfo,
)


# =============================================================================
# SMAST
# =============================================================================

class SmastMetaclass(DeclarativeMeta):
    # noinspection PyInitNewSignature
    def __init__(cls: Type['Smast'],
                 name: str,
                 bases: Tuple[Type, ...],
                 classdict: Dict[str, Any]) -> None:
        add_multiple_columns(
            cls, "q", 1, cls.NQUESTIONS, CharColType,
            pv=['Y', 'N'],
            comment_fmt="Q{n}: {s} (Y or N)",
            comment_strings=[
                "believe you are a normal drinker",
                "near relative worries/complains",
                "feel guilty",
                "friends/relative think you are a normal drinker",
                "stop when you want to",
                "ever attended Alcoholics Anonymous",
                "problems with close relative",
                "trouble at work",
                "neglected obligations for >=2 days",
                "sought help",
                "hospitalized",
                "arrested for drink-driving",
                "arrested for other drunken behaviour",
            ]
        )
        super().__init__(name, bases, classdict)


class Smast(TaskHasPatientMixin, Task,
            metaclass=SmastMetaclass):
    __tablename__ = "smast"
    shortname = "SMAST"
    longname = "Short Michigan Alcohol Screening Test"
    provides_trackers = True

    NQUESTIONS = 13
    TASK_FIELDS = strseq("q", 1, NQUESTIONS)

    def get_trackers(self, req: CamcopsRequest) -> List[TrackerInfo]:
        return [TrackerInfo(
            value=self.total_score(),
            plot_label="SMAST total score",
            axis_label="Total score (out of {})".format(self.NQUESTIONS),
            axis_min=-0.5,
            axis_max=self.NQUESTIONS + 0.5,
            horizontal_lines=[
                2.5,
                1.5,
            ],
            horizontal_labels=[
                TrackerLabel(4, self.wxstring(req, "problem_probable")),
                TrackerLabel(2, self.wxstring(req, "problem_possible")),
                TrackerLabel(0.75, self.wxstring(req, "problem_unlikely")),
            ]
        )]

    def get_clinical_text(self, req: CamcopsRequest) -> List[CtvInfo]:
        if not self.is_complete():
            return CTV_INCOMPLETE
        return [CtvInfo(
            content="SMAST total score {}/{} ({})".format(
                self.total_score(), self.NQUESTIONS, self.likelihood(req))
        )]

    def get_summaries(self, req: CamcopsRequest) -> List[SummaryElement]:
        return self.standard_task_summary_fields() + [
            SummaryElement(
                name="total",
                coltype=Integer(),
                value=self.total_score(),
                comment="Total score (/{})".format(self.NQUESTIONS)),
            SummaryElement(
                name="likelihood",
                coltype=SummaryCategoryColType,
                value=self.likelihood(req),
                comment="Likelihood of problem"),
        ]

    def is_complete(self) -> bool:
        return (
            self.are_all_fields_complete(self.TASK_FIELDS) and
            self.field_contents_valid()
        )

    def get_score(self, q: int) -> int:
        yes = "Y"
        value = getattr(self, "q" + str(q))
        if value is None:
            return 0
        if q == 1 or q == 4 or q == 5:
            return 0 if value == yes else 1
        else:
            return 1 if value == yes else 0

    def total_score(self) -> int:
        total = 0
        for q in range(1, self.NQUESTIONS + 1):
            total += self.get_score(q)
        return total

    def likelihood(self, req: CamcopsRequest) -> str:
        score = self.total_score()
        if score >= 3:
            return self.wxstring(req, "problem_probable")
        elif score >= 2:
            return self.wxstring(req, "problem_possible")
        else:
            return self.wxstring(req, "problem_unlikely")

    def get_task_html(self, req: CamcopsRequest) -> str:
        score = self.total_score()
        likelihood = self.likelihood(req)
        main_dict = {
            None: None,
            "Y": req.wappstring("yes"),
            "N": req.wappstring("no")
        }
        q_a = ""
        for q in range(1, self.NQUESTIONS + 1):
            q_a += tr(
                self.wxstring(req, "q" + str(q)),
                answer(get_from_dict(main_dict, getattr(self, "q" + str(q)))) +
                " — " + str(self.get_score(q))
            )
        h = """
            <div class="{CssClass.SUMMARY}">
                <table class="{CssClass.SUMMARY}">
                    {tr_is_complete}
                    {total_score}
                    {problem_likelihood}
                </table>
            </div>
            <table class="{CssClass.TASKDETAIL}">
                <tr>
                    <th width="80%">Question</th>
                    <th width="20%">Answer</th>
                </tr>
                {q_a}
            </table>
            <div class="{CssClass.FOOTNOTES}">
                [1] Total score ≥3 probable, ≥2 possible, 0–1 unlikely.
            </div>
        """.format(
            CssClass=CssClass,
            tr_is_complete=self.get_is_complete_tr(req),
            total_score=tr(
                req.wappstring("total_score"),
                answer(score) + " / {}".format(self.NQUESTIONS)
            ),
            problem_likelihood=tr_qa(
                self.wxstring(req, "problem_likelihood") + " <sup>[1]</sup>",
                likelihood
            ),
            q_a=q_a,
        )
        return h
